File: js/gtsdb.js
/**
 * GTSDB JavaScript/Node.js client over TCP with JSON and binary protocols.
 *
 *   const { GTSDBClient } = require('gtsdb-drivers');
 *   const client = new GTSDBClient('localhost', 5555);
 *   await client.connect();
 *   await client.auth('your-token');
 *   await client.write('sensor1', 42.5);
 *   const points = await client.readBinary('sensor1', 100);
 */

const net = require('net');

class GTSDBClient {
  constructor(host, port) {
    this.host = host;
    this.port = port;
    this.socket = new net.Socket();
    this.buffer = '';
    this.pendingResolve = null;
    this.pendingReject = null;
    this.subCallbacks = new Map();
    this.connected = false;
    this._dataHandler = null;
  }

  connect() {
    return new Promise((resolve, reject) => {
      this.socket.connect(this.port, this.host, () => {
        this.connected = true;
        resolve();
      });
      this.socket.on('error', reject);
      this.socket.on('data', (chunk) => this._onData(chunk));
    });
  }

  close() {
    return new Promise((resolve) => {
      this.socket.end(() => {
        this.connected = false;
        resolve();
      });
    });
  }

  /** Send a JSON request and return the raw response line. */
  _send(req) {
    return new Promise((resolve, reject) => {
      this.pendingResolve = (data) => {
        try { resolve(JSON.parse(data)); } catch (e) { resolve(data); }
      };
      this.pendingReject = reject;
      this.socket.write(JSON.stringify(req) + '\n');
    });
  }

  /** Send a JSON request and read a length-prefixed binary response frame. */
  _sendBinary(req) {
    return new Promise((resolve, reject) => {
      this._binaryResolve = resolve;
      this._binaryReject = reject;
      this._binaryState = 0; // 0: reading length, 1: reading data
      this._binaryLen = 0;
      this._binaryBuf = null;
      this._binaryOff = 0;
      this.socket.write(JSON.stringify(req) + '\n');
    });
  }

  _onData(chunk) {
    // If we're in binary read mode, handle binary framing
    if (this._binaryResolve) {
      this._handleBinaryChunk(chunk);
      return;
    }

    // JSON line mode
    this.buffer += chunk.toString();
    const lines = this.buffer.split('\n');
    this.buffer = lines.pop(); // keep incomplete line

    for (const line of lines) {
      if (!line.trim()) continue;
      try {
        const msg = JSON.parse(line);
        // Check if this is a subscription update
        if (msg.data && msg.data.key && this.subCallbacks.has(msg.data.key)) {
          for (const cb of this.subCallbacks.get(msg.data.key)) {
            cb(msg.data);
          }
          continue;
        }
        // Otherwise, resolve pending request
        if (this.pendingResolve) {
          this.pendingResolve(msg);
          this.pendingResolve = null;
          this.pendingReject = null;
        }
      } catch (e) {
        if (this.pendingReject) {
          this.pendingReject(e);
          this.pendingResolve = null;
          this.pendingReject = null;
        }
      }
    }
  }

  /** Handle binary protocol framing. */
  _handleBinaryChunk(chunk) {
    let off = 0;
    const data = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);

    while (off < data.length) {
      if (this._binaryState === 0) {
        // Reading 4-byte length prefix
        const need = 4 - this._binaryOff;
        const avail = data.length - off;
        const take = Math.min(need, avail);

        if (this._binaryLenBuf === undefined) this._binaryLenBuf = Buffer.alloc(4);
        data.copy(this._binaryLenBuf, this._binaryOff, off, off + take);
        this._binaryOff += take;
        off += take;

        if (this._binaryOff === 4) {
          this._binaryLen = this._binaryLenBuf.readUInt32BE(0);
          this._binaryState = 1;
          this._binaryOff = 0;
          this._binaryBuf = Buffer.alloc(this._binaryLen);
          this._binaryLenBuf = undefined;
          if (this._binaryLen === 0) {
            this._finishBinary();
            return;
          }
        }
      } else {
        // Reading frame data
        const need = this._binaryLen - this._binaryOff;
        const avail = data.length - off;
        const take = Math.min(need, avail);

        data.copy(this._binaryBuf, this._binaryOff, off, off + take);
        this._binaryOff += take;
        off += take;

        if (this._binaryOff === this._binaryLen) {
          this._finishBinary();
          return;
        }
      }
    }
  }

  _finishBinary() {
    const resolve = this._binaryResolve;
    const buf = this._binaryBuf;
    this._binaryResolve = null;
    this._binaryReject = null;
    this._binaryBuf = null;
    resolve(this._parseBinaryMulti(buf));
  }

  /** Parse a binary frame into a map of key → points. */
  _parseBinaryMulti(buf) {
    if (buf.length < 4) return {};
    let off = 0;
    const result = {};
    const numKeys = buf.readUInt32BE(off);
    off += 4;

    for (let i = 0; i < numKeys; i++) {
      if (off + 2 > buf.length) break;
      const keyLen = buf.readUInt16BE(off);
      off += 2;
      if (off + keyLen > buf.length) break;
      const key = buf.toString('utf8', off, off + keyLen);
      off += keyLen;
      if (off + 4 > buf.length) break;
      const pointCount = buf.readUInt32BE(off);
      off += 4;

      const pts = [];
      for (let j = 0; j < pointCount; j++) {
        if (off + 16 > buf.length) break;
        const ts = Number(buf.readBigInt64BE(off));
        const val = buf.readDoubleBE(off + 8);
        off += 16;
        pts.push({ key, timestamp: ts, value: val });
      }
      result[key] = pts;
    }
    return result;
  }

  // ── Public API ──────────────────────────────────

  /** Authenticate with a token. */
  async auth(token) {
    const r = await this._send({ operation: 'auth', key: token });
    if (!r.success) throw new Error('GTSDB auth failed: ' + r.message);
  }

  /** Write a single data point. */
  async write(key, value, timestamp = 0) {
    const req = { operation: 'write', key, write: { value } };
    if (timestamp > 0) req.write.timestamp = timestamp;
    return this._send(req);
  }

  /** Read data points in a time range (JSON). */
  async read(key, startTs, endTs, downsample = 0) {
    const r = await this._send({
      operation: 'read', key,
      read: { start_timestamp: startTs, end_timestamp: endTs, downsampling: downsample },
    });
    return r.data || [];
  }

  /** Read the last N data points (JSON). */
  async readLast(key, n) {
    const r = await this._send({ operation: 'read', key, read: { lastx: n } });
    return r.data || [];
  }

  /** Read the last N data points using the fast binary protocol. */
  async readBinary(key, n) {
    const multi = await this._sendBinary({
      operation: 'read', key,
      read: { lastx: n },
      response_format: 'binary',
    });
    return multi[key] || [];
  }

  /** Read from multiple keys using binary protocol. */
  async multiReadBinary(keys, lastX) {
    return this._sendBinary({
      operation: 'multi-read', keys,
      read: { lastx: lastX },
      response_format: 'binary',
    });
  }

  /** Get all known sensor IDs. */
  async ids() {
    const r = await this._send({ operation: 'ids' });
    return r.data || [];
  }

  /** Subscribe to real-time updates for a key. */
  async subscribe(key, callback) {
    if (!this.subCallbacks.has(key)) this.subCallbacks.set(key, []);
    this.subCallbacks.get(key).push(callback);
    return this._send({ operation: 'subscribe', key });
  }

  /** Unsubscribe from updates for a key. */
  async unsubscribe(key) {
    this.subCallbacks.delete(key);
    return this._send({ operation: 'unsubscribe', key });
  }

  /** Batch write multiple data points. */
  async batchWrite(points) {
    return this._send({
      operation: 'batch-write',
      points: points.map(p => ({ key: p.key, value: p.value, timestamp: p.timestamp || 0 })),
    });
  }
}

module.exports = { GTSDBClient };

File: go/gtsdb.go
// Package gtsdb provides a Go client for GTSDB over TCP with JSON and binary protocols.
//
//	client, _ := gtsdb.Connect("localhost:5555")
//	client.Auth("your-token")
//	client.Write("sensor1", 42.5)
//	points, _ := client.ReadBinary("sensor1", 100)
package gtsdb

import (
	"bufio"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net"
	"sync"
)

// DataPoint represents a single time-series data point.
type DataPoint struct {
	Key       string  `json:"key"`
	Timestamp int64   `json:"timestamp"`
	Value     float64 `json:"value"`
}

// Client is a GTSDB TCP client.
type Client struct {
	conn   net.Conn
	reader *bufio.Reader
	mu     sync.Mutex
}

// Connect establishes a TCP connection to a GTSDB server.
func Connect(addr string) (*Client, error) {
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return nil, fmt.Errorf("gtsdb: dial %s: %w", addr, err)
	}
	return &Client{
		conn:   conn,
		reader: bufio.NewReader(conn),
	}, nil
}

// Close closes the connection.
func (c *Client) Close() error {
	return c.conn.Close()
}

// send writes a JSON-encoded request and returns the raw JSON response line.
func (c *Client) send(req interface{}) ([]byte, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	data, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("gtsdb: marshal: %w", err)
	}
	data = append(data, '\n')
	if _, err := c.conn.Write(data); err != nil {
		return nil, fmt.Errorf("gtsdb: write: %w", err)
	}
	return c.reader.ReadBytes('\n')
}

// sendBinary sends a JSON request and reads a length-prefixed binary response frame.
func (c *Client) sendBinary(req interface{}) ([]byte, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	data, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("gtsdb: marshal: %w", err)
	}
	data = append(data, '\n')
	if _, err := c.conn.Write(data); err != nil {
		return nil, fmt.Errorf("gtsdb: write: %w", err)
	}
	return readBinaryFrame(c.reader)
}

// readBinaryFrame reads a 4-byte length-prefixed binary frame.
func readBinaryFrame(r io.Reader) ([]byte, error) {
	var lenBuf [4]byte
	if _, err := io.ReadFull(r, lenBuf[:]); err != nil {
		return nil, fmt.Errorf("gtsdb: read frame len: %w", err)
	}
	n := binary.BigEndian.Uint32(lenBuf[:])
	if n == 0 {
		return nil, nil
	}
	data := make([]byte, n)
	if _, err := io.ReadFull(r, data); err != nil {
		return nil, fmt.Errorf("gtsdb: read frame data: %w", err)
	}
	return data, nil
}

// parseBinaryMulti parses a binary multi-data frame into a map of key → points.
func parseBinaryMulti(data []byte) (map[string][]DataPoint, error) {
	if len(data) < 4 {
		return nil, nil
	}
	result := make(map[string][]DataPoint)
	off := 0
	numKeys := binary.BigEndian.Uint32(data[off:])
	off += 4
	for i := uint32(0); i < numKeys; i++ {
		if off+2 > len(data) {
			break
		}
		keyLen := binary.BigEndian.Uint16(data[off:])
		off += 2
		if off+int(keyLen) > len(data) {
			break
		}
		key := string(data[off : off+int(keyLen)])
		off += int(keyLen)
		if off+4 > len(data) {
			break
		}
		pointCount := binary.BigEndian.Uint32(data[off:])
		off += 4
		pts := make([]DataPoint, pointCount)
		for j := uint32(0); j < pointCount; j++ {
			if off+16 > len(data) {
				break
			}
			pts[j] = DataPoint{
				Key:       key,
				Timestamp: int64(binary.BigEndian.Uint64(data[off:])),
				Value:     math.Float64frombits(binary.BigEndian.Uint64(data[off+8:])),
			}
			off += 16
		}
		result[key] = pts
	}
	return result, nil
}

// parseBinaryKey parses a single-key binary frame.
func parseBinaryKey(data []byte) ([]DataPoint, error) {
	parsed, err := parseBinaryMulti(data)
	if err != nil {
		return nil, err
	}
	for _, pts := range parsed {
		return pts, nil
	}
	return nil, nil
}

// Auth authenticates with a token.
func (c *Client) Auth(token string) error {
	resp, err := c.send(map[string]string{
		"operation": "auth",
		"key":       token,
	})
	if err != nil {
		return err
	}
	var r jsonResponse
	json.Unmarshal(resp, &r)
	if !r.Success {
		return fmt.Errorf("gtsdb: auth failed: %s", r.Message)
	}
	return nil
}

// Write sends a single data point. Timestamp is set server-side if 0.
func (c *Client) Write(key string, value float64) error {
	return c.WriteAt(key, value, 0)
}

// WriteAt sends a single data point with a specific timestamp.
func (c *Client) WriteAt(key string, value float64, timestamp int64) error {
	req := map[string]interface{}{
		"operation": "write",
		"key":       key,
		"write": map[string]interface{}{
			"value": value,
		},
	}
	if timestamp > 0 {
		req["write"].(map[string]interface{})["timestamp"] = timestamp
	}
	_, err := c.send(req)
	return err
}

// Read reads data points in a time range.
func (c *Client) Read(key string, startTs, endTs int64, downsample int) ([]DataPoint, error) {
	resp, err := c.send(map[string]interface{}{
		"operation": "read",
		"key":       key,
		"read": map[string]interface{}{
			"start_timestamp": startTs,
			"end_timestamp":   endTs,
			"downsampling":    downsample,
		},
	})
	if err != nil {
		return nil, err
	}
	var r struct {
		Data []DataPoint `json:"data"`
	}
	json.Unmarshal(resp, &r)
	return r.Data, nil
}

// ReadLast reads the last N data points for a key.
func (c *Client) ReadLast(key string, n int) ([]DataPoint, error) {
	resp, err := c.send(map[string]interface{}{
		"operation": "read",
		"key":       key,
		"read":      map[string]interface{}{"lastx": n},
	})
	if err != nil {
		return nil, err
	}
	var r struct {
		Data []DataPoint `json:"data"`
	}
	json.Unmarshal(resp, &r)
	return r.Data, nil
}

// ReadBinary reads the last N points using the fast binary protocol.
func (c *Client) ReadBinary(key string, n int) ([]DataPoint, error) {
	data, err := c.sendBinary(map[string]interface{}{
		"operation":       "read",
		"key":             key,
		"read":            map[string]interface{}{"lastx": n},
		"response_format": "binary",
	})
	if err != nil {
		return nil, err
	}
	return parseBinaryKey(data)
}

// ReadLastBinary is an alias for ReadBinary.
func (c *Client) ReadLastBinary(key string, n int) ([]DataPoint, error) {
	return c.ReadBinary(key, n)
}

// MultiRead reads from multiple keys in one request using JSON.
func (c *Client) MultiRead(keys []string, lastX int) (map[string][]DataPoint, error) {
	resp, err := c.send(map[string]interface{}{
		"operation": "multi-read",
		"keys":      keys,
		"read":      map[string]interface{}{"lastx": lastX},
	})
	if err != nil {
		return nil, err
	}
	var r struct {
		MultiData map[string][]DataPoint `json:"multi_data"`
	}
	json.Unmarshal(resp, &r)
	return r.MultiData, nil
}

// MultiReadBinary reads from multiple keys using the fast binary protocol.
func (c *Client) MultiReadBinary(keys []string, lastX int) (map[string][]DataPoint, error) {
	data, err := c.sendBinary(map[string]interface{}{
		"operation":       "multi-read",
		"keys":            keys,
		"read":            map[string]interface{}{"lastx": lastX},
		"response_format": "binary",
	})
	if err != nil {
		return nil, err
	}
	return parseBinaryMulti(data)
}

// IDs returns all known sensor keys.
func (c *Client) IDs() ([]string, error) {
	resp, err := c.send(map[string]string{"operation": "ids"})
	if err != nil {
		return nil, err
	}
	var r struct {
		Data []string `json:"data"`
	}
	json.Unmarshal(resp, &r)
	return r.Data, nil
}

// Subscribe subscribes to real-time updates for a key.
// The callback receives data points as they arrive.
func (c *Client) Subscribe(key string, cb func(DataPoint)) error {
	_, err := c.send(map[string]string{
		"operation": "subscribe",
		"key":       key,
	})
	if err != nil {
		return err
	}
	go func() {
		for {
			resp, err := c.reader.ReadBytes('\n')
			if err != nil {
				return
			}
			var r struct {
				Success bool      `json:"success"`
				Data    DataPoint `json:"data"`
			}
			if json.Unmarshal(resp, &r) == nil && r.Success {
				cb(r.Data)
			}
		}
	}()
	return nil
}

// Unsubscribe stops receiving updates for a key.
func (c *Client) Unsubscribe(key string) error {
	_, err := c.send(map[string]string{
		"operation": "unsubscribe",
		"key":       key,
	})
	return err
}

// BatchWrite writes multiple data points in one request.
func (c *Client) BatchWrite(points []DataPoint) error {
	batch := make([]map[string]interface{}, len(points))
	for i, p := range points {
		batch[i] = map[string]interface{}{
			"key":       p.Key,
			"value":     p.Value,
			"timestamp": p.Timestamp,
		}
	}
	_, err := c.send(map[string]interface{}{
		"operation": "batch-write",
		"points":    batch,
	})
	return err
}

type jsonResponse struct {
	Success bool   `json:"success"`
	Message string `json:"message"`
}
